torch_struct/distributions.py

```diff
@@ -504,4 +484,26 @@ def argmax(self):
 
     @lazy_property
     def entropy(self):
-        pass
+        r"""
+        Compute entropy efficiently using arc-factorization property.
+
+        Algorithm derivation:
+        ..math::
+        {{
+            \begin{align}
+            H[p] &= E_{p(T)}[-\log p(T)]\\
+            &= -E_{p(T)}\big[ \log [\frac{1}{Z} \prod\limits_{(i,j) \in T} \exp\{\phi_{i,j}\}] \big]\\
+            &= -E_{p(T)}\big[  \sum\limits_{(i,j) \in T} \phi_{i,j} - \log Z \big]\\
+            &= \log Z -E_{p(T)}\big[\sum\limits_{(i,j) \in A} 1\{(i,j) \in T\} \phi_{i,j}\big]\\
+            &= \log Z - \sum\limits_{(i,j) \in A} p\big((i,j) \in T\big) \phi_{i,j}
+            \end{align}
+        }}
+
+        Returns:
+            entropy (*batch_shape)
+        """
+        logZ = self.partition
```

Comment:
@justinchiu points out that Instead of grouping this with non-proj we should just have it be the default entropy function for all of the models to materialize the marginals. (nothing non-projective specific).

Reply:
Yes, it's a general property of all exponential family / sum-product models.  See also this Twitter discussion https://twitter.com/RanZmigrod/status/1300832956701970434?s=20